Fixes issue 2503: Add network tests for BlockRewards and fix bug

hewison-chris opened the pull request:
Adding Network Unit tests to verify block rewards are not missing.

hewison-chris committed:
Add network tests for BlockRewards

Verify the ammounts in the `Coinbase` payouts to `Commons Budget` and
`Validators` for several scenarios to ensure payouts are as defined in
the `Bosagora Whitepaper`. Now with test that has a slashed node and
also a node only signing half of the blocks.

hewison-chris commented:
So far the basic tests are creating blocks with the expected rewards. I will need to add some more scenarios to replicate the issue seen in Production. Maybe the loss came from having a slashed validator.

Geod24 commented on source/agora/test/BlockRewards.d:
```diff
@@ -0,0 +1,85 @@
+    });
+}
+
+/// TODO: test with a validator slashed
+
+/// TODO: test with some missing signatures
```
I guess this is still a draft then ? Especially in light of #2503

hewison-chris replied:
Yes, I will add the tests today should then fix #2503.

hewison-chris commented:
I have added a test where one validator only signs half the blocks. It fails at the moment but will try and fix by Monday.

hewison-chris committed:
Correct height used for block rewards

hewison-chris committed:
Correct documentation comments for Block Rewards

hewison-chris committed:
Use Amount percentage function

hewison-chris committed:
Only calculate reduced payout when required

hewison-chris committed:
Increase max blocks allowed to be fetched

Increase from 20 to 24 so that more than one test validator cycle can be
fetched which can be useful when testing.

hewison-chris committed:
Trivial documentation correction

kchulj commented:
Looks green?
